File: javascript/src/socketd/transport/core/Session.ts

import {Entity} from "./Entity";
import {Message} from "./Message";
import {Consumer} from "../../utils/Consumer";

export interface Session {
    isValid(): boolean

    remoteAddress(): object

    localAddress(): object

    handshake(): object

    param(name: string): string

    paramOrDefault(name: string, value: string): string

    path(): string

    pathNew(pathNew: string): void

    // @ts-ignore
    attrMap(): Map<string, object>

    attr<T>(name: string): T

    attrOrDefault<T>(name: string, def: T): T

    attrSet<T>(name: string, value: T): void

    sessionId(): string

    reconnect(): void

    sendPing(): void

    send(topic: string, entity: Entity): void

    sendAndRequest(topic: string, entity: Entity): Entity

    sendAndSubscribe(topic: string, entity: Entity, consumer: Consumer<Entity>): void

    reply(from: Message, entity: Entity): void

    replyEnd(from: Message, entity: Entity): void
}

File: go/socketd/transport/core/session.go
package core

import (
	"net"
	"socketd/transport/core/message"
)

type Session interface {
	LocalAddress() net.Addr
	RemoteAddress() net.Addr
	Handshake() *Handshake
	SessionId() string

	Param(name string) string                      // 握手参数
	ParamOrDefault(name string, def string) string // 获取握手参数/默认值
	Path() string                                  // 握手路径
	PathNew(path string)                           // 设置新握手路径

	AttrHas(name string) bool               // 是否有该属性
	AttrOrDefault(name string, def any) any // 获取属性值/默认值
	Attr(name string) any                   // 获取属性值
	AttrPut(name string, value any)         // 放置属性值

	Reconnect() error                                           // 重新连接(手动)
	SendPing() error                                            // 手动发送一个ping包
	SendAlarm(from *message.Frame, alarm string) error          // 发送一个告警
	Reply(from *message.Frame, entity *message.Entity) error    // 回复
	ReplyEnd(from *message.Frame, entity *message.Entity) error // 最后一个回复，回复即断开
}
